cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java

```diff
@@ -101,37 +100,62 @@ public void put(Job job) {
   public Job transitState(Job job, Job.State state) {
     // assert that the job state transition is valid
     job.getState().checkTransition(state);
-    SimpleJob newJob = new SimpleJob(job.getSchedule(), job.getCreationTime(), job.getNotifications(), state);
+    Job newJob = new SimpleJob(job.getSchedule(), job.getCreationTime(), job.getNotifications(), state);
     put(newJob);
     return newJob;
   }
 
   @Override
   public void addNotification(ProgramSchedule schedule, Notification notification) {
-    boolean existingJobModified = false;
+    boolean jobExists = false;
     try (CloseableIterator<Job> jobs = getJobsForSchedule(schedule.getScheduleId())) {
       while (jobs.hasNext()) {
+        jobExists = true;
         Job job = jobs.next();
-        if (job.getState() != Job.State.PENDING_LAUNCH) {
-          // check if its PENDING_LAUNCH. If so, create a new Job to avoid the chance that the job-launching
-          // process has conflict
+        if (job.getState() == Job.State.PENDING_TRIGGER) {
+          // only update the job's notifications if it is in PENDING_TRIGGER, so as to avoid conflict with the
+          // ConstraintCheckerService
           addNotification(job, notification);
-          existingJobModified = true;
+          break;
         }
       }
     }
-    // if no existing job was modified with the new notification, add a new job with the first notification
-    if (!existingJobModified) {
-      Job job = new SimpleJob(schedule, System.currentTimeMillis(),
-                              Lists.newArrayList(notification), Job.State.PENDING_TRIGGER);
-      put(job);
+    // if no job exists for the scheduleId, add a new job with the first notification
+    if (!jobExists) {
+      addNotification(new SimpleJob(schedule, System.currentTimeMillis(),
```

Comment:
I think it'll be simpler if you just add the job with the correct state in here:

```java
List<Notification> notifications = Arrays.asList(notification);
Job.State jobState = isTriggerSatisfied(schedule.getTrigger(), notifications) 
                      ? Job.State.PENDING_CONSTRAINT : Job.State.PENDING_TRIGGER;
put(new SimpleJob(schedule, System.currentTimeMillis(), notificats, jobState));
```

You can also turn these three lines into a private method `addNewJob`.